bin/tx-prover/src/proxy/mod.rs

```diff
@@ -0,0 +1,193 @@
+};
+use pingora_core::{upstreams::peer::HttpPeer, Result};
+use pingora_limits::rate::Rate;
+use pingora_proxy::{ProxyHttp, Session};
+use tokio::sync::RwLock;
+use tracing::error;
+
+/// Timeout for the requests
+const TIMEOUT_SECS: Option<Duration> = Some(Duration::from_secs(100));
+/// Maximum number of items per queue
+const MAX_QUEUE_ITEMS: usize = 10;
+
+/// Load balancer that uses a round robin strategy
+pub struct LB(Arc<LoadBalancer<RoundRobin>>);
+
+impl LB {
+    pub fn new(upstreams: LoadBalancer<RoundRobin>) -> Self {
+        Self(Arc::new(upstreams))
+    }
+}
+
+/// Rate limiter
+static RATE_LIMITER: Lazy<Rate> = Lazy::new(|| Rate::new(Duration::from_secs(1)));
+
+/// Maximum amount of request per second per client
+static MAX_REQ_PER_SEC: isize = 5;
+
+/// Shared state. It is a map of backends to a vector of request IDs
+static QUEUES: Lazy<RwLock<HashMap<Backend, Vec<String>>>> =
+    Lazy::new(|| RwLock::new(HashMap::new()));
+
+#[async_trait]
+impl ProxyHttp for LB {
+    type CTX = ();
+
+    fn new_ctx(&self) {}
+
+    async fn upstream_peer(
+        &self,
+        session: &mut Session,
+        _ctx: &mut Self::CTX,
+    ) -> Result<Box<HttpPeer>> {
+        let upstream = self.0.select(b"", 256).unwrap();
+
+        // read request ID from headers
+        let request_id = session
+            .get_header("X-Request-ID")
+            .expect("Request ID not found")
+            .to_str()
+            .expect("Invalid header value");
+
+        {
+            let mut ctx_guard = QUEUES.write().await;
+            let backend_queue = ctx_guard.entry(upstream.clone()).or_insert_with(Vec::new);
+
+            // Limit queue length to MAX_QUEUE_ITEMS requests
+            if backend_queue.len() >= MAX_QUEUE_ITEMS {
+                panic!("Too many requests in the queue");
+                // return Err(Box::new("Too many requests in the queue".into()));
+            }
+
+            backend_queue.push(request_id.to_string());
+        }
+
+        // Wait for the request to be at the front of the queue
+        loop {
+            // We use a new scope for each iteration to release the lock
+            {
+                let ctx_guard = QUEUES.read().await;
+                if let Some(backend_queue) = ctx_guard.get(&upstream) {
+                    if backend_queue[0] == request_id {
+                        break;
+                    }
+                } else {
+                    // TODO: replace this panic with an error
+                    panic!("Upstream not found");
+                }
+            }
+            tokio::time::sleep(Duration::from_millis(100)).await;
+        }
+
+        // Set SNI
+        let mut http_peer = HttpPeer::new(upstream, false, "".to_string());
+        let peer_opts = http_peer.get_mut_peer_options().unwrap();
+
+        // Timeout settings
+        peer_opts.total_connection_timeout = TIMEOUT_SECS;
+        peer_opts.connection_timeout = TIMEOUT_SECS;
```

Comment:
I think for the connection timeout we might want to keep it lower than the transaction proving timeout, perhaps something within 5-10 seconds